python/ray/_private/job_manager/job_manager.py

```diff
@@ -152,21 +172,39 @@ def run(self, cmd: str):
             stdout_path, stderr_path = self._log_client.get_log_file_paths(
                 self._job_id)
 
-            exit_code = exec_cmd_logs_to_file(cmd, stdout_path, stderr_path)
-        finally:
-            # 3) Once command finishes, update status to SUCCEEDED or FAILED.
-            if exit_code == 0:
-                self._status = JobStatus.SUCCEEDED
+            task, child_proc = await self._exec_cmd(cmd, stdout_path,
+                                                    stderr_path)
+
+            finished, _ = await asyncio.wait(
+                [task, self._stop_event.wait()], return_when=FIRST_COMPLETED)
+
+            if self._stop_event.is_set():
+                self._status_client.put_status(self._job_id, JobStatus.STOPPED)
+                task.cancel()
+                child_proc.kill()
             else:
-                self._status = JobStatus.FAILED
-            self._status_client.put_status(self._job_id, self.get_status())
+                # Child process finished execution
+                assert len(
+                    finished) == 1, "Should have only one coroutine done"
+                for task in finished:
+                    return_code = task.result()
+                    if return_code == 0:
+                        self._status_client.put_status(self._job_id,
+                                                       JobStatus.SUCCEEDED)
+                    else:
+                        self._status_client.put_status(self._job_id,
+                                                       JobStatus.FAILED)
+        except Exception:
+            logger.error(traceback.format_exc())
+        finally:
+            # clean up actor after tasks are finished
             ray.actor.exit_actor()
 
     def get_status(self) -> JobStatus:
```

Comment:
we don't need this as a public method anymore, right?

Reply:
yep ! addressed